Dot grid

adamgerhant opened the pull request:
Closes [#1704](https://github.com/GraphiteEditor/Graphite/issues/1704)

Adds a dot type grid option. Snapping and spacing is the same as the rectangular grid.

Changes:
- Adds `grid_overlay_dot`function to draw circles using the overlay context
- Adds `GridType::Dot` with similar implementation to `GridType::Rectangle`
- UI changes for selecting the Dot option
- Replaces matches! macro for selecting index to a match statement
- Adds circle function to [utility_types.rs](https://github.com/GraphiteEditor/Graphite/pull/1709/files#diff-fc1f68d21385d829e2a2a8eadca8f208e900ee9c3262a58fcc79ae5ff4251733)

![dots](https://github.com/GraphiteEditor/Graphite/assets/116332429/863896c9-dd5a-4a08-a81a-7329afad5101)

adamgerhant commented:
Added grid color select to close #1703. 

Changes:
- Adds support for prefixed hexed codes
- Adds grid_color field to GridSnapping 
- Color selector UI

Note: blocked by #1710 since initial color is not correct.

0HyperCube commented on editor/src/messages/portfolio/document/overlays/grid_overlays.rs:
```diff
@@ -135,9 +171,16 @@ pub fn overlay_options(grid: &GridSnapping) -> Vec<LayoutGroup> {
 				RadioEntryData::new("isometric")
 					.label("Isometric")
 					.on_update(update_val(grid, |grid, _| grid.grid_type = GridType::ISOMETRIC)),
+				RadioEntryData::new("dot")
+					.label("Dot")
+					.on_update(update_val(grid, |grid, _| grid.grid_type = GridType::DOT)),
```
The dot grid works well, but I think that it should not be displayed as a seperate grid type (as it is merely an alternative rendering method for the rectangle grid). Probably a check box if the rectangle grid enabled would work better.

0HyperCube commented on node-graph/gcore/src/raster/color.rs:
````diff
@@ -831,6 +860,26 @@ impl Color {
 		Some(Color::from_rgba8_srgb(r, g, b, a))
 	}
 
+	/// Creates a color from a 8-character RGBA hex string (with a # prefix).
+	///
+	/// # Examples
+	/// ```
+	/// use graphene_core::raster::color::Color;
+	/// let color = Color::from_rgba_str_prefixed("#7C67FA61").unwrap();
+	/// ```
+	pub fn from_rgba_str_prefixed(color_str: &str) -> Option<Color> {
+		if color_str.len() != 9 {
+			return None;
+		}
+   		let color_str = &color_str[1..];
+		let r = u8::from_str_radix(&color_str[0..2], 16).ok()?;
+		let g = u8::from_str_radix(&color_str[2..4], 16).ok()?;
+		let b = u8::from_str_radix(&color_str[4..6], 16).ok()?;
+		let a = u8::from_str_radix(&color_str[6..8], 16).ok()?;
+
+		Some(Color::from_rgba8_srgb(r, g, b, a))
+	}
````
Please reuse the existing function above to reduce code duplication. e.g. `colour_string.strip_prefix("#").and_then(|value|Color::from_rgba_str(value)`. This will also ensure that it is actually a `#`.

0HyperCube commented on node-graph/gcore/src/raster/color.rs:
````diff
@@ -847,6 +896,23 @@ impl Color {
 		Some(Color::from_rgb8_srgb(r, g, b))
 	}
 
+	/// Creates a color from a 6-character RGB hex string (with a # prefix).
+	/// ```
+	/// use graphene_core::raster::color::Color;
+	/// let color = Color::from_rgb_str_prefixed("7C67FA").unwrap();
+	/// ```
+	pub fn from_rgb_str_prefixed(color_str: &str) -> Option<Color> {
+		if color_str.len() != 7 {
+			return None;
+		}
+		let color_str = &color_str[1..];
+		let r = u8::from_str_radix(&color_str[0..2], 16).ok()?;
+		let g = u8::from_str_radix(&color_str[2..4], 16).ok()?;
+		let b = u8::from_str_radix(&color_str[4..6], 16).ok()?;
+
+		Some(Color::from_rgb8_srgb(r, g, b))
+	}
````
Again `colour_string.strip_prefix("#").and_then(|value|Color::from_rgb_str(value)` will suffice.

0HyperCube reviewed (changes requested):
Thanks for your effort so far.

adamgerhant commented:
I believe all changes should have been made. I removed the Dot grid type, and instead added a Boolean field that tracks whether the grid should be displayed with dots. This also makes it easy to add a dot grid for the Isometric type, if there are plans for that. Just remove the condition for displaying the checkbox, create the display function, and use it accordingly.

I removed the prefixed color methods and used concatenation and remove_prefix accordingly. Currently this PR is blocked by the Color bug, since it starts at `#999999` instead of `#cccccc`. An easy fix could be to replace the hex constant with a Color.

Keavon commented:
!build

Keavon commented:
Square 1px pixel grid aligned (no antialiasing) dots would already be the desired appearance. Just be aware that a dotted line might drift if the ratio is a non-integer fraction between the document space and viewport space, so some cleverness might be required to enable the avoidance of antialiasing by staying perfectly grid aligned. Maybe several lines per row/column with different screenspace pixel spacing intervals as the error accumulates between the non-integer fraction ratio and the ratio you need to maintain screenspace grid alignment.

Keavon commented:
I'm putting this back to draft status till you have time again to deal with the performance suggestions in my previous comment. But don't focus on this till your other PR is done, since that's the much higher priority feature. Thanks :)

Keavon commented:
!build
